design/WhitePaper.md

```diff
@@ -0,0 +1,713 @@
+```
+actor {
+  var list : List<Nat> = null;
+  var len : Nat = 0;
+  public putLogged(x : Nat) : async () {
+    let := len + 1;
+    let y = await other.process(x);
+    list := List.push(y, list);
+    log("old head: " # Nat.toText(List.head(List.head(list))));
+  }
+}
+```
+Here, if the logging attempt fails because the original list was empty, then the mutation of `list` will be abandoned, but not the mutation of `len`, because that happened before the `await`, and thus has already been committed to the actor's state.
+
+There is no easy general solution to this problem – that would require the implementation of distributed transactions on the IC level, which are notoriously hard and expensive. But again it might be desirable to warn against accidental state mutations before an await, unless the programmer declares intent explicitly – and thereby, presumably, was aware of the implications.
+
+
+### Primitive Types
+
+Primitive types in Motoko still lack basic functionality.
+
+#### Number Conversions
+
+Currently, converting between different number types requires going through `Nat` or `Int` in most cases:
+```
+func nat8ToNat16(n : Nat8) : Nat16 { Nat18.fromNat(Nat8.toNat(n)) }
+```
+Obviously, that's both inconvenient and expensive. The base library should directly support the complete conversion matrix for all number types (nats, ints, and floats).
+
+
+#### Blobs
+
+Blobs are essentially arrays of bytes, but they currently lack basic array functionality, like random access. That ought to be added in the same way as for arrays.
+
+
+#### Text
+
+All text strings are in Unicode in Motoko. But there is no proper support for processing Unicode. For example, there are no Unicode-aware comparison operators, and no Unicode character classification or conversion functions.
+
+Adding these requires including Unicode libraries and tables into the generated Wasm code, which can easily be several MB in size. Presumably, supporting Unicode would hence depend on some tree shaking ability of the compiler.
+
+Another shortcoming in text processing is the inability to efficiently extract substrings. Motoko provides iterators over strings, but they cannot be used as arguments to other operations, because their type is structural. A solution to this is needed, e.g., by extending text iterator objects in some way. Details tbd.
+
+Finally, it would be desirable to have extended syntax for multi-line text literals, ideally with interpolation, perhaps based on a generic toText function (see [shared generics](#shared-generics)).
+
+
+#### Floats
+
+While Motoko is the first blockchain language to feature floating point numbers, their functionality is still limited to the primitives that Wasm provides. The base library should include common math functionality like transcendental functions. Since these have to be implemented in software (as part of the runtime), any imprecision in such computations remains deterministic.
```

Comment:
```suggestion
While Motoko is the first blockchain language to feature floating point numbers, their functionality is still limited to the primitives that Wasm provides. The base library should include common math functionality like transcendental functions. Since these have to be implemented in software (as part of the runtime), any imprecision in such computations remains deterministic.
```
I think we do already - are we missing some? https://github.com/dfinity/motoko-base/blob/master/src/Float.mo

Reply:
Ugh, indeed, I forgot. Sorry Gabor :) Removed.

Reply:
I guess there's the question if we want to support float32 since some AI algorithms don't mind the loss of precision.